File: 2023/07.10/송기훈/트리의지름.java
package baekjoon.gold.four;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class 트리의지름 {

    static class Node {
        int point;
        int distance;

        public Node(int point, int distance) {
            this.point = point;
            this.distance = distance;
        }
    }

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int N;
    static ArrayList<ArrayList<Node>> connections;
    static boolean[] visited;
    static int maxValue, maxIndex;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());

        connections = new ArrayList<>();

        for (int i = 0; i < N + 1; i++) {
            connections.add(new ArrayList<Node>());
        }

        for (int i = 0; i < N - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int parent = Integer.parseInt(st.nextToken());
            int child = Integer.parseInt(st.nextToken());
            int distance = Integer.parseInt(st.nextToken());

            connections.get(parent).add(new Node(child, distance));
            connections.get(child).add(new Node(parent, distance));
        }

        maxValue = Integer.MIN_VALUE;
        maxIndex = -1;
        // 어떤 경로든 1을 거쳐가야 하므로
        // 1에서 제일 먼 곳 탐색하면 답을 구할 수 있다.
        visited = new boolean[N + 1];
        visited[1] = true;
        dfs(1, 0);

        // maxIndex에서 제일 먼 곳 탐색
        maxValue = Integer.MIN_VALUE;
        visited = new boolean[N + 1];
        visited[maxIndex] = true;
        dfs(maxIndex, 0);

        System.out.println(maxValue);

    }


    static void dfs(int index, int distance) {
        if (maxValue < distance) {
            maxValue = distance;
            maxIndex = index;
        }

        ArrayList<Node> nodeList = connections.get(index);

        for (Node node : nodeList) {
            if (!visited[node.point]) {
                visited[node.point] = true;
                dfs(node.point, distance + node.distance);
            }
        }

    }

}

File: 2023/07.10/조수연/트리의_지름.kt
import java.io.*
import java.util.StringTokenizer
import kotlin.math.*

lateinit var tree: ArrayList<ArrayList<Pair<Int, Int>>>
lateinit var checked: BooleanArray // 리프 노드 체크 배열
lateinit var visited: BooleanArray // 중복 방문 체크 배열
lateinit var finished: BooleanArray // 리프 노드 중복 방문 체크 배열
var answer = 0

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val n = readLine().toInt()

    tree = ArrayList()

    for (i in 0..n) {
        tree.add(ArrayList())
    }

    checked = BooleanArray(n + 1){true}
    finished = BooleanArray(n + 1)

    repeat(n - 1) {
        val token = StringTokenizer(readLine())
        val parent = token.nextToken().toInt()
        val child = token.nextToken().toInt()
        val weight = token.nextToken().toInt()

        tree[parent].add(Pair(child, weight))
        tree[child].add(Pair(parent, weight))

        checked[parent] = false
    }

    val leaves = ArrayList<Int>()

    for (i in checked.indices) {
        if (checked[i]) leaves.add(i)
    }

    for (i in leaves.indices) {
        visited = BooleanArray(n + 1)
        checked[leaves[i]] = false
        maxDiameter(leaves[i],0)
        finished[leaves[i]] = true
    }

    bw.write("$answer")

    bw.flush()
    bw.close()
}

// dfs 접근
fun maxDiameter(start: Int, weight:Int) {

    visited[start] = true

    for (next in tree[start]) {
        if (!visited[next.first] && !finished[next.first]) {
            maxDiameter(next.first,weight + next.second)
        }
    }

    answer = max(answer,weight)
    return
}